### PR TITLE
[Inference]Adapt temperature processing logic

### PR DESCRIPTION
**Pull request by yuehuayingxueluo**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Adapt temperature processing logic, and tests have been conducted on Baichuan2-13B and llama3 accordingly.

**pytest:**
image

**Inference Test:**
***Baichuan2-13B:***
temperature(1.0)
image
temperature(0.7)
image
temperature(0.5)
image

***llama3***
temperature(1.0)
image
temperature(0.7)
image
temperature(0.5)
image

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/policies/llama.py [hunk 1]
```diff
@@ -141,9 +141,10 @@ def module_policy(self) -> Dict[Union[str, nn.Module], ModulePolicyDescription]:
             assert (
                 self.model.config.num_attention_heads % self.shard_config.tensor_parallel_size == 0
             ), f"The number of attention heads must be divisible by tensor parallel size."
-            assert (
-                self.model.config.num_key_value_heads % self.shard_config.tensor_parallel_size == 0
-            ), f"The number of key_value heads must be divisible by tensor parallel size."
+            if hasattr(self.model.config, "num_key_value_heads"):
+                assert (
+                    self.model.config.num_key_value_heads % self.shard_config.tensor_parallel_size == 0
+                ), f"The number of key_value heads must be divisible by tensor parallel size."
```
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply colossalai/shardformer/policies/llama.py:147]: Yes, Baichuan doesn't has num_key_value_heads.

File: colossalai/inference/logit_processors.py [hunk 2]
```diff
@@ -32,6 +48,9 @@ def top_p_logit_processor(logits, top_p: float):
     """
     top_p logit processor
     """
+    if top_p == 1.0:
+        return logits
+
```
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply colossalai/inference/logit_processors.py:53]: I think when top_p == 1.0, we can skip the top computation logic and avoid extra computation.

File: colossalai/inference/logit_processors.py [hunk 3]
```diff
@@ -17,6 +17,22 @@ def register(func):
...
+@register_logit_processor("temperature")
+def temperature_logit_process(logits, temperature: float):
+    """
+    apply temperature scaling.
+    """
+    if isinstance(temperature, float) and temperature > 0 and temperature <= 1.0:
+        if temperature == 1.0:
+            return logits
+        else:
+            return logits / temperature
```
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply colossalai/inference/logit_processors.py:29]: OK

File: colossalai/inference/logit_processors.py [hunk 4]
```diff
@@ -17,6 +17,22 @@ def register(func):
...
+    else:
+        raise ValueError(
+            f"'temperature={temperature}' should be a strictly positive float, less than or equal to 1.0 and greater than 0."
+        )
```
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply colossalai/inference/logit_processors.py:33]: OK

**Review by yuanheng-zhao — APPROVED** [review]
LGTM

**Pull request merged**